### PR TITLE
Add more TypeConverter tests

### PR DESCRIPTION
**Pull request by hughbe**
- Enhance the test base to cover more behaviour that should be uniform across all converters
- Use in GuidConverter and DateTimeConverter
- Remove checks for if the `ctor` is null in DateTimeConverter when converting to an InstanceDescriptor as they should always exist

### CODE REVIEW

File: src/System.ComponentModel.TypeConverter/tests/TypeConverterTestBase.cs [hunk 1]
```diff
@@ -358,5 +393,10 @@ public string GetSerializedString()
                 }
             }
         }
+
+        private class CustomToString
+        {
+            public override string ToString() => "CustomToString";
```
Comment (maryamariyan) [inline src/System.ComponentModel.TypeConverter/tests/TypeConverterTestBase.cs]: nit: nameof(CustomToString)

File: src/System.ComponentModel.TypeConverter/tests/GuidConverterTests.cs [hunk 2]
```diff
@@ -10,89 +10,42 @@
...
-            InstanceDescriptor descriptor = Assert.IsType<InstanceDescriptor>(converter.ConvertTo(value, typeof(InstanceDescriptor)));
-            ConstructorInfo constructor = Assert.IsAssignableFrom<ConstructorInfo>(descriptor.MemberInfo);
-            Assert.Equal(new Type[] { typeof(string) }, constructor.GetParameters().Select(p => p.ParameterType));
-            Assert.Equal(new object[] { "30da92c0-23e8-42a0-ae7c-734a0e5d2782" }, descriptor.Arguments);
+            yield return ConvertTest.Valid(new Guid(0x30da92c0, 0x23e8, 0x42a0, 0xae, 0x7c, 0x73, 0x4a, 0x0e, 0x5d, 0x27, 0x82), "30da92c0-23e8-42a0-ae7c-734a0e5d2782");
+            yield return ConvertTest.Valid(
+                new Guid(0x30da92c0, 0x23e8, 0x42a0, 0xae, 0x7c, 0x73, 0x4a, 0x0e, 0x5d, 0x27, 0x82),
+                new InstanceDescriptor(
+                    typeof(Guid).GetConstructor(new Type[] { typeof(string) }),
+                    new object[] { "30da92c0-23e8-42a0-ae7c-734a0e5d2782" }
```
Comment (maryamariyan) [inline src/System.ComponentModel.TypeConverter/tests/GuidConverterTests.cs]: nit: 30da92c0-23e8-42a0-ae7c-734a0e5d2782 is used multiple times in the file, you can save it as a variable.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReferenceConverter.cs [hunk 3]
```diff
@@ -130,23 +124,25 @@ public override object ConvertTo(ITypeDescriptorContext context, CultureInfo cul
...
 
             if (context != null)
             {
-                var list = new List<object> { null };
+                components = new List<object> { null };
 
                 // Try the reference service first.
-                IReferenceService refSvc = (IReferenceService)context.GetService(typeof(IReferenceService));
-                if (refSvc != null)
+                if (context.GetService(typeof(IReferenceService)) is IReferenceService refSvc)
```
Comment (hughbe) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReferenceConverter.cs]: Fix InvalidCastException

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReferenceConverter.cs [hunk 4]
```diff
@@ -73,29 +72,24 @@ public override object ConvertFrom(ITypeDescriptorContext context, CultureInfo c
...
-            }
-
             if (destinationType == typeof(string))
             {
                 if (value != null)
                 {
                     // Try the reference service first.
-                    IReferenceService refSvc = (IReferenceService)context?.GetService(typeof(IReferenceService));
-                    if (refSvc != null)
+                    if (context?.GetService(typeof(IReferenceService)) is IReferenceService refSvc)
```
Comment (hughbe) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReferenceConverter.cs]: Fix InvalidCastException

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReferenceConverter.cs [hunk 5]
```diff
@@ -45,15 +46,13 @@ public override bool CanConvertFrom(ITypeDescriptorContext context, Type sourceT
...
+            if (value is string text)
             {
-                string text = ((string)value).Trim();
-
+                text = text.Trim();
                 if (!string.Equals(text, s_none) && context != null)
                 {
                     // Try the reference service first.
-                    IReferenceService refSvc = (IReferenceService)context.GetService(typeof(IReferenceService));
-                    if (refSvc != null)
```
Comment (hughbe) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReferenceConverter.cs]: Fix InvalidCastException

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReferenceConverter.cs [hunk 6]
```diff
@@ -130,23 +124,25 @@ public override object ConvertTo(ITypeDescriptorContext context, CultureInfo cul
         /// </summary>
         public override StandardValuesCollection GetStandardValues(ITypeDescriptorContext context)
         {
-            object[] components = null;
+            List<object> components = null;
```
Comment (hughbe) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReferenceConverter.cs]: Reduce allocations

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReferenceConverter.cs [hunk 7]
```diff
@@ -156,20 +152,17 @@ public override StandardValuesCollection GetStandardValues(ITypeDescriptorContex
                     if (cont != null)
                     {
                         ComponentCollection objs = cont.Components;
-
                         foreach (IComponent obj in objs)
                         {
-                            if (obj != null && _type.IsInstanceOfType(obj) &&
-                                IsValueAllowed(context, obj))
+                            if (obj != null && _type != null && _type.IsInstanceOfType(obj) && IsValueAllowed(context, obj))
```
Comment (hughbe) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReferenceConverter.cs]: Fix NullReferenceException

**Comment by maryamariyan** [comment]
Thanks @hughbe for the changes.

**Pull request merged**